tests/__init__.py

```diff
@@ -1,40 +1,85 @@
-from typing import Any, Dict, Optional
+# Copyright 2022 New Vector Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sqlite3
+from asyncio import Future
+from typing import Any, Awaitable, Callable, Optional, Tuple, TypeVar
+from unittest.mock import Mock
 
-import attr
-from mock import Mock
-from synapse.module_api import ModuleApi
+from synapse.module_api import JsonDict, ModuleApi
 
 from tchap_red_list import RedListManager
 
+RV = TypeVar("RV")
+TV = TypeVar("TV")
 
-@attr.s(auto_attribs=True)
-class MockEvent:
-    """Mocks an event. Only exposes properties the module uses."""
-    sender: str
-    type: str
-    content: Dict[str, Any]
-    room_id: str = "!someroom"
-    state_key: Optional[str] = None
 
-    def is_state(self) -> bool:
-        """Checks if the event is a state event by checking if it has a state key."""
-        return self.state_key is not None
+class SQLiteStore:
+    """In-memory SQLite store. We can't just use a run_db_interaction function that opens
+    its own connection, since we need to use the same connection for all queries in a
+    test.
+    """
 
-    @property
-    def membership(self) -> str:
-        """Extracts the membership from the event. Should only be called on an event
-        that's a membership event, and will raise a KeyError otherwise.
-        """
-        membership: str = self.content["membership"]
-        return membership
+    def __init__(self) -> None:
+        self.conn = sqlite3.connect(":memory:")
 
+    async def run_db_interaction(
+        self, desc: str, f: Callable[..., RV], *args: Any, **kwargs: Any
```

Comment:
If you're feeling brave, `ParamSpec` can be used here:
```py
from typing_extensions import ParamSpec

P = ParamSpec("P")

async def run_db_interaction(
        self, desc: str, f: Callable[P, RV], *args: P.args, **kwargs: P.kwargs
```
`typing_extensions` will need to be added to dev dependencies.

With that said I don't think this buys us anything in terms of type checking, so it may not be worth it.

Reply:
Yeah, plus Tchap isn't running Python 3.10+